iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -122,11 +115,18 @@ private void InitializeClient()
             }
         }
 
-        private void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
+        // It is not good practice to have async void methods, however, DeviceClient.SetConnectionStatusChangesHandler() event handler signature has a void return type.
+        // As a result, any operation within this block will be executed unmonitored on another thread.
+        // To prevent multi-threaded synchronization issues, the async method InitializeClientAsync being called in here first grabs a lock
+        // before attempting to initailize or dispose the device client instance.
+        private async void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
```

Comment:
this is definitely a bad pattern, but that is because of the way `ConnectionStatusChangeHandler` is defined. We have locks within the async method `InitializeClientAsync`, to ensure that there are no multi-threaded issues.